vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/server/ClientConnector.java

```diff
@@ -0,0 +1,130 @@
+package com.vaadin.legacy.server;/*
+                                 *
+                                 *  * #%L
+                                 *  * Vaadin Legacy Components for Vaadin 21
+                                 *  * %%
+                                 *  * Copyright (C) 2021 Vaadin Ltd
+                                 *  * %%
+                                 *  * This program is available under Commercial Vaadin Developer License
+                                 *  * 4.0 (CVDLv4).
+                                 *  *
+                                 *  * See the file license.html distributed with this software for more
+                                 *  * information about licensing.
+                                 *  *
+                                 *  * For the full License, see <https://vaadin.com/license/cvdl-4.0>.
+                                 *  * #L%
+                                 *
+                                 */
+
+import com.vaadin.flow.component.Component;
+import com.vaadin.flow.component.ComponentEvent;
+import com.vaadin.flow.component.ComponentEventListener;
+import com.vaadin.flow.shared.Registration;
+import com.vaadin.flow.component.ComponentUtil;
+
+public interface ClientConnector {
+
+    /**
+     * Add a listener for connector attach events.
+     *
+     * @param listener
+     * @return Registration for unregistering the listener
+     */
+    Registration addAttachListener(AttachListener listener);
+
+    /**
+     * Add a listener for connector detach events.
+     *
+     * @param listener
+     * @return Registration for unregistering the listener
+     */
+    Registration addDetachListener(DetachListener listener);
+
+    /**
+     * Notifies the connector that it is connected to a VaadinSession (and
+     * therefore also to a UI).
+     * <p>
+     * The caller of this method is
+     * {@link ComponentUtil#onComponentAttach(Component, boolean)}} if the
+     * parent is itself already attached to the session. If not, it's called
+     * when the parent is attached to the session. This method is always called
+     * before the connector's data is sent to the client-side for the first
+     * time.
+     * </p>
+     *
+     * <p>
+     * The attachment logic is implemented in {@link AbstractClientConnector}.
+     * </p>
+     */
+    void attach();
+
+    /**
+     * Notifies the connector that it is detached from its VaadinSession.
+     *
+     * </p>
+     */
+    void detach();
+
+    /**
+     * Checks if the connector is attached to a VaadinSession.
+     *
+     * @return true if the connector is attached to a session, false otherwise
+     */
+    boolean isAttached();
+
+    /**
+     * Interface for listening {@link AttachEvent connector attach events}.
+     *
+     */
+    @FunctionalInterface
+    interface AttachListener
+            extends ComponentEventListener<ClientConnector.AttachEvent> {
+    }
```

Comment:
This is not actually compatible with the legacy API when not used as an lambda, which is probably the most typical use case. But backwards compatibility for all cases can be achieved by
```suggestion
        @Override
        default void onComponentEvent(ClientConnector.AttachEvent event) {
            this.attach(event);
        }

        /**
         * Called when a AttachListener is notified of a AttachEvent.
         *
         * @param event
         *            The attach event that was fired.
         */
         void attach(ClientConnector.AttachEvent event);
    }
```